### PR TITLE
Multiple repos load them

### LINKED ISSUE #62: Repos: Support multiple repos through and through
[`multiple-repos`](https://github.com/iFixit/pulldasher/compare/multiple-repos)

At the moment, pulldasher only supports a single repository. Let's update it to support more than one. We'll still only have one dashboard, but it may have pulls from across multiple repos.

A few record types contain a `repo` column, but nearly all of them will need it (and need it filled in)

### PR DESCRIPTION
**Pull request by MooreGuy**
After the pull to store repository data is merged and working, then this will start reading from it.

Allows pulldasher to read from multiple repositories.

~~Built on top of #81~~ (merged)

Closes iFixit/ifixit#18725
Closes #62

### CODE REVIEW

File: lib/git-manager.js [hunk 1]
```diff
@@ -326,6 +329,27 @@ function params(apiParams) {
 }
 
 /**
+ * Returns a function that uses the search parameters to add the "repository"
+ * property to all the results. When we ask for a list of open issues from the
+ * API for a particular repo, those results don't have references to the repo
+ * we asked about, so we have to inject them to normalize the structure of the
+ * object.
```
Comment (djmetzle) [inline lib/git-manager.js]: Please adjust this comment

File: models/db_issue.js [hunk 2]
```diff
@@ -1,9 +1,12 @@
 var utils = require('../lib/utils');
 var db = require('../lib/db');
-var utils = require('../lib/utils');
```
Comment (djmetzle) [inline models/db_issue.js]: Seems odd..

File: models/pull.js [hunk 3]
```diff
@@ -189,6 +158,42 @@ Pull.parseBody = function(body) {
...
+   var data = {
+      number: data.number,
+      state: data.state,
+      title: data.title,
+      body: data.body,
+      created_at: utils.fromDateString(data.created_at),
+      updated_at: utils.fromDateString(data.updated_at),
+      closed_at: utils.fromDateString(data.closed_at),
+      merged_at: utils.fromDateString(data.merged_at),
+      difficulty: data.difficulty,
```
Comment (djmetzle) [inline models/pull.js]: `difficulty`? That's a thing?

File: lib/git-manager.js (hunk 1 above)
Comment (danielbeardsley) [inline lib/git-manager.js]: Adjust it for what? More context would help when making suggestions.

File: migrations/0007-add-repo-to-all-tables.sql [hunk 4]
```diff
@@ -1,3 +1,6 @@
+ALTER TABLE `comments`
+ADD COLUMN `repo` VARCHAR(255) NULL BEFORE `repo_name`;
```
Comment (djmetzle) [inline migrations/0007-add-repo-to-all-tables.sql]: `BEFORE`?

File: lib/git-manager.js (hunk 1 above)
Comment (djmetzle) [inline lib/git-manager.js]: Its a four line run-on?
> so we have to inject them
Inject what?

I get the intent of the comment, but it's really opaque.

File: lib/git-manager.js [hunk 5]
```diff
@@ -259,7 +259,10 @@ function getAllPages(currentPage, allResults) {
  * get*Issues() commands and thus has been augmented with the
  * issue.repo property.
  */
-function getLabelsFromEvents(events, ghIssue, repoName) {
+function getLabelsFromEvents(events, ghIssue) {
+   // until we remove this entirely, let's jsut get it the easy way.
+   var repoName = ghIssue.repository.full_name.split('/')[1];
```
Comment (djmetzle) [inline lib/git-manager.js]: Hacks!

File: lib/db-manager.js [hunk 6]
```diff
@@ -432,12 +432,12 @@ var dbManager = module.exports = {
...
     */
-   deleteLabels: function deleteLabels(number) {
+   deleteLabels: function deleteLabels(repo, number) {
       debug('Calling deleteLabels for pull #%s', number);
       var q_delete = 'DELETE FROM pull_labels WHERE number = ? ' +
-       'AND repo_name = ?';
-
-      return db.query(q_delete, [number, config.repo.name]);
+       'AND repo = ?';
+name
```
Comment (djmetzle) [inline lib/db-manager.js]: This looks like an error..

File: lib/db-manager.js [hunk 7]
```diff
@@ -436,7 +436,6 @@ var dbManager = module.exports = {
       debug('Calling deleteLabels for pull #%s', number);
       var q_delete = 'DELETE FROM pull_labels WHERE number = ? ' +
        'AND repo = ?';
-name
```
Comment (djmetzle) [inline lib/db-manager.js]: Yerp.

File: lib/pull-queue.js [hunk 8]
```diff
@@ -15,9 +15,9 @@ _.extend(PullQueue.prototype, {
     */
    markPullAsDirty: function(repo, number) {
       if (this.queuePaused) {
-         this.dirtyPulls[number] = true;
+         this.dirtyPulls[repo + "/" + number] = true;
```
Comment (djmetzle) [inline lib/pull-queue.js]: This seems to beg for using an object instead of a string.

**Comment by djmetzle** [comment]
Looks good! :+1:
CR :bookmark:

File: lib/git-manager.js (hunk 1 above)
Comment (danielbeardsley) [inline lib/git-manager.js]: How about:

> Returns a function that uses the search parameters to add the "repository"
property to an array of result objects (or a single result).
> 
> When we ask for a list of open issues from the
> API for a particular repo, those results don't have references to the repo
> we asked about. We inject these references to normalize the structure
of the object used to construct pulls.

Comment (djmetzle) [inline lib/git-manager.js]: Much better.

File: migrations/0008-add-repo-primary-key.sql [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ALTER TABLE commit_statuses ADD PRIMARY KEY (repo, commit);
+---
+ALTER TABLE issues DROP PRIMARY KEY;
+ALTER TABLE issues ADD PRIMARY KEY (repo, number);
+---
+ALTER TABLE pull_labels DROP PRIMARY KEY;
+ALTER TABLE pull_labels ADD PRIMARY KEY (repo, number, title);
+---
+ALTER TABLE pulls DROP PRIMARY KEY;
+ALTER TABLE pulls ADD PRIMARY KEY (repo, number);
```
Comment (danielbeardsley) [inline migrations/0008-add-repo-primary-key.sql]: This change isn't reflected in the schema.sql below.

File: schema.sql [hunk 10]
```diff
@@ -90,7 +102,8 @@ CREATE TABLE IF NOT EXISTS `pull_labels` (
 
 /*!40101 SET @saved_cs_client     = @@character_set_client */;
 /*!40101 SET character_set_client = utf8 */;
-CREATE TABLE IF NOT EXISTS `pull_signatures` (
+CREATE TABLE `pull_signatures` (
+  `repo` varchar(255) DEFAULT NULL,
```
Comment (danielbeardsley) [inline schema.sql]: I think we *should* add this column to the `pull_signatures_number` key though.

**Comment by MooreGuy** [comment]
dev_block :+1: On @danielbeardsley 's pull being merged in: #81 since this will be faster to CR once it is merged.

**Comment by MooreGuy** [comment]
un_dev_block :+1: It's merged!

**Comment by djmetzle** [comment]
CR 💥

File: config.example.js [hunk 11]
```diff
@@ -41,13 +41,13 @@ module.exports = {
...
       secret: "secret for signing session cookies"
    },
-   repo: {
-      // The GitHub username of the person or organization owning the repo you
-      // want to monitor.
-     owner:       'repo owner',
+
+   // List of repositories for pulldasher to watch.
+   repo: [
      // The name of the GitHub repo you want to monitor.
```
Comment (andrewjpiro) [inline config.example.js]: This comment doesn't make sense anymore

Comment (MooreGuy) [inline config.example.js]: db037a6

File: app.js [hunk 12]
```diff
@@ -55,18 +55,20 @@ authManager.setupRoutes(app);
...
+   dbManager.getOpenPulls(repo).then(function(pulls) {
+      pullQueue.pause();
+      pulls.forEach(function(pull) {
+         pullManager.updatePull(pull);
+      });
+      pullQueue.resume();
+   })
+   // Get the most recent version of each pull from the API
+   .then(function () {
+      return refresh.openPulls();
```
Comment (danielj41) [inline app.js]: Would it be better for this to wait for all of the `getOpenPulls` promises?

File: lib/git-manager.js [hunk 13]
```diff
@@ -359,36 +363,64 @@ function addRepo(params) {
...
+ *
+ * The function should take a repository name, and return an array of values.
+ */
+function forEachRepo(singleRepoLambda, args) {
+   // default value
+   args = args || [];
+   args.unshift(null);
+   var allRepoMap = function(currentRepo) {
+      args[0] = currentRepo;
+      return singleRepoLambda.call(this, args);
```
Comment (danielj41) [inline lib/git-manager.js]: Is this supposed to be `apply`?

File: lib/refresh.js [hunk 14]
```diff
@@ -12,42 +12,42 @@ var pullQueue  = new NotifyQueue();
...
       debug("refresh issue %s", number);
-      return gitManager.getIssue(number)
+      return gitManager.getIssue(repo, number)
       .then(pushOnQueue(issueQueue));
    },
 
    allIssues: function refreshOpenIssues() {
       debug("refresh all issues");
-      return gitManager.getAllIssues()
+      return gitManager.forEachRepo(gitManager.getAllIssues())
```
Comment (danielj41) [inline lib/refresh.js]: Should this be `getAllIssues` without the `()`? Unless `getAllIssues` returns a function?

File: lib/refresh.js [hunk 15]
```diff
@@ -12,42 +12,42 @@ var pullQueue  = new NotifyQueue();
...
       debug("refresh pull %s", number);
-      return gitManager.getPull(number)
+      return gitManager.getPull(repo, number)
       .then(pushOnQueue(pullQueue));
    },
 
    allPulls: function refreshOpenPulls() {
       debug("refresh all pull");
-      return gitManager.getAllPulls()
+      return gitManager.forEachRepo(gitManager.getAllPulls)
```
Comment (danielj41) [inline lib/refresh.js]: I assume it's not necessary to `.bind(gitManager)` on these functions?

File: lib/db-manager.js [hunk 16]
```diff
@@ -445,3 +446,12 @@ var dbManager = module.exports = {
    // @TODO closeStalePulls:
 
 };
+
+/**
+ * Return a copy of the specified array with nulls removed
+ */
+function filterNulls(array) {
+   return array.filter(function(element) {
+      return element;
```
Comment (danielj41) [inline lib/db-manager.js]: Kind of pedantic, but I feel like this should return `element !== null` or the function should be called `filterFalsy`.

**Comment by djmetzle** [comment]
CR :+1: Looks good,
But does it work @MooreGuy ?

File: lib/pull-queue.js (hunk 8 above)
Comment (danielj41) [inline lib/pull-queue.js]: It matches `npm`'s `package.json` syntax for github repos :)

File: lib/git-manager.js (hunk 13 above)
Comment (danielj41) [inline lib/git-manager.js]: Solved in a later commit, please ignore.

**Comment by danielj41** [comment]
Nice work!

CR 👍 but I'd like responses to my comments.

File: app.js (hunk 12 above)
Comment (danielbeardsley) [inline app.js]: Not exactly sure what you mean here. 

We have:
```
getOpenPulls().then(
   Do synchronous stuff
).then(
   return refresh.openPulls()
).done()
```

Maybe you didn't know that `updatePull()` is synchronous?

Comment (danielj41) [inline app.js]: Yeah, you're right. Ignore this then :)

File: lib/db-manager.js [hunk 17]
```diff
@@ -298,11 +297,12 @@ var dbManager = module.exports = {
     * Returns a promise which resolves to an array of Pull objects for the
     * given pull numbers built from data stored in the database.
     */
-   getPulls: function(pullNumbers, repo) {
+   getPulls: function(repo, pullNumbers) {
```
Comment (danielbeardsley) [inline lib/db-manager.js]: Looks like we don't even use this function... huh.

File: lib/pull-queue.js [hunk 18]
```diff
@@ -13,11 +13,11 @@ _.extend(PullQueue.prototype, {
     * Adds the pull number to the queue if it is paused; otherwise, updates
```
Comment (MooreGuy) [inline lib/pull-queue.js]: I stopped here. @danielbeardsley

**Comment by danielj41** [comment]
CR 👍

File: lib/pull-queue.js (hunk 18 above)
Comment (danielbeardsley) [inline lib/pull-queue.js]: I made it all the way through and made a few more commits (see below commits)

File: controllers/githubHooks.js [hunk 19]
```diff
@@ -28,7 +28,7 @@ var HooksController = {
       }
 
       // Begin the webhook decoding
-      var body = JSON.parse(req.body.payload);
```
Comment (danielj41) [inline controllers/githubHooks.js]: I assume it gets parsed by `app.use(bodyParser.json())` in `app.js`?

I don't know how this code worked in the first place, though.

**Comment by danielj41** [comment]
CR 👍

File: controllers/githubHooks.js (hunk 19 above)
Comment (MooreGuy) [inline controllers/githubHooks.js]: I'm not really sure, I was going to try and investigate, but I'm pretty unfamiliar with this.

Comment (MooreGuy) [inline controllers/githubHooks.js]: Maybe @danielbeardsley or @copperwall or @andyg0808 have any ideas?

**Comment by MooreGuy** [comment]
I'm looking into why when you sign off on a pull it doesn't get updated in the UI. dev_block :+1:

File: controllers/githubHooks.js (hunk 19 above)
Comment (danielbeardsley) [inline controllers/githubHooks.js]: See my comment on this commit.

**Comment by MooreGuy** [comment]
Okay, that fixes the problem that I was searching for. I'm now going to update the webhook decoding, and that will hopefully be the last thing.

**Comment by MooreGuy** [comment]
un_dev_block :+1:

**Comment by danielj41** [comment]
CR 👍

**Comment by djmetzle** [comment]
CR :open_umbrella:

**Comment by danielj41** [comment]
CR 👍

**Comment by djmetzle** [comment]
CR 🕓

**Comment by djmetzle** [comment]
CR :+1: again

**Comment by MooreGuy** [comment]
There is a bug where multiple pull signatures are showing up from the same person when they should have been invalidated. dev_block 👍

**Comment by danielj41** [comment]
CR 👍

**Comment by djmetzle** [comment]
CR 👍

**Comment by andrewjpiro** [comment]
Looks great to me!
Tested:
 - CR/QA
 - dev_block/deploy_block
 - committing to invalidate signoffs
 - Updating milestones
 - Adding QA labels
 - Watched leaderboards while I was doing this / counted for accuracy
 
Minor things found:
 - We added a # in front of pull numbers unnecessarily

Also it thinks I'm Danny. But I heard that's not an issue.
QA :its-happening:

**Comment by MooreGuy** [comment]
Cool, I'm going to merge and deploy

**Pull request merged**